state.go

```diff
@@ -27,76 +31,804 @@ type State struct {
+			Application: &ApplicationData{
+				Data: data,
+			},
+		},
+	}
+
+	pt.sign(s.groupContext(), s.IdentityPriv, s.scheme)
+	return s.encrypt(pt)
 }
 
 func (s *State) unprotect(ct *MLSCiphertext) ([]byte, error) {
-	// TODO
-	return nil, nil
+	pt, err := s.decrypt(ct)
+	if err != nil {
+		return nil, err
+	}
+
+	sigPubKey := s.Tree.GetCredential(pt.Sender).PublicKey()
+	if !pt.verify(s.groupContext(), sigPubKey, s.scheme) {
+		return nil, fmt.Errorf("invalid message signature")
+	}
+
+	if pt.Content.Type() != ContentTypeApplication {
+		return nil, fmt.Errorf("unprotect attempted on non-application message")
+	}
+	return pt.Content.Application.Data, nil
+}
+
+func senderDataAAD(gid []byte, epoch Epoch, contentType ContentType, nonce []byte) []byte {
+	s := NewWriteStream()
+	err := s.Write(struct {
+		GroupId         []byte `tls:"head=1"`
+		Epoch           Epoch
+		ContentType     ContentType
+		SenderDataNonce []byte `tls:"head=1"`
+	}{
+		GroupId:         gid,
+		Epoch:           epoch,
+		ContentType:     contentType,
+		SenderDataNonce: nonce,
+	})
+
+	if err != nil {
+		return nil
+	}
+
+	return s.Data()
+}
+
+func contentAAD(gid []byte, epoch Epoch,
+	contentType ContentType, authenticatedData []byte,
+	nonce []byte, encSenderData []byte) []byte {
+
+	s := NewWriteStream()
+	err := s.Write(struct {
+		GroupId             []byte `tls:"head=1"`
+		Epoch               Epoch
+		ContentType         ContentType
+		AuthenticatedData   []byte `tls:"head=4"`
+		SenderDataNonce     []byte `tls:"head=1"`
+		EncryptedSenderData []byte `tls:"head=1"`
+	}{
+		GroupId:             gid,
+		Epoch:               epoch,
+		ContentType:         contentType,
+		AuthenticatedData:   authenticatedData,
+		SenderDataNonce:     nonce,
+		EncryptedSenderData: encSenderData,
+	})
+
+	if err != nil {
+		return nil
+	}
+	return s.Data()
+}
+
+func (s State) clone() *State {
+	// Note: all the slice/map copy operations below on state are mere
+	// reference copies.
+	cloned := new(State)
+	cloned.CipherSuite = s.CipherSuite
+	cloned.GroupID = append(s.GroupID[:0:0], s.GroupID...)
+	cloned.Epoch = s.Epoch
+	cloned.Tree = *s.Tree.clone()
+	cloned.ConfirmedTranscriptHash = append(s.ConfirmedTranscriptHash[:0:0], s.ConfirmedTranscriptHash...)
+	cloned.InterimTranscriptHash = append(s.InterimTranscriptHash[:0:0], s.InterimTranscriptHash...)
+	cloned.Keys = s.Keys
+	cloned.Index = s.Index
+	cloned.IdentityPriv = s.IdentityPriv
+	cloned.scheme = s.scheme
+	if s.PendingProposals != nil {
+		cloned.PendingProposals = append(s.PendingProposals[:0:0], s.PendingProposals...)
+	}
+	cloned.UpdateSecrets = s.UpdateSecrets
+	cloned.processedProposals = s.processedProposals
+	return cloned
+}
+
+// Compare the public aspects of two nodes
+func (s State) Equals(o State) bool {
+
+	if s.Epoch != o.Epoch || !bytes.Equal(s.GroupID, o.GroupID) || s.CipherSuite != o.CipherSuite {
+		return false
+	}
+
+	return s.Tree.Equals(&o.Tree) && reflect.DeepEqual(s.Keys, o.Keys)
```

Comment:
Comparing the key schedules like this will be a little brittle, since the state of the key schedule changes depending on what messages have been sent/received.  Also, the key schedule is not a "public aspect".